### PR TITLE
Add firmware and documentation for NF18, Temperature and Humidity Monitor.

### PR DESCRIPTION
**Pull request by haydenroche5**

### CODE REVIEW

File: 18-temperature-and-humidity-monitor/firmware/README.md [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+5. Press and hold the button labeled "BOOT" on the Swan, and, while holding BOOT, press and release the button labeled "RST", and finally release the BOOT button. The Swan is now ready to be flashed.
+6. Click "Upload" under Project Tasks > bw_swan_r5 > General. You should see "SUCCESS" in the terminal output pane.
+
+## Operation
+
+At this point, the Swan firmware should be running and you should start to see events show up on Notehub under your project's "Events" view. There are 3 important Notefiles.
+
+### `data.qo`
+
+By default, the Swan will publish temperature and humidity data from the BME280 to this file every 3 minutes. Ex:
```
Comment (m-mcgowan) [inline 18-temperature-and-humidity-monitor/firmware/README.md:76]: Is `Ex:` a normal way to write Example? I would prefer something like

> For example:

> A typical event looks like this:

>  "E.g." is "For example"

Probably too terse and formal for our style, but just mentioning it. (Although having said that, it does feature on dev.blues.io.)

File: 18-temperature-and-humidity-monitor/firmware/include/README [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+place, and programs that include the header file will automatically use the
+new version when next recompiled. The header file eliminates the labor of
+finding and changing all the copies as well as the risk that a failure to
+find one copy will result in inconsistencies within a program.
+
+In C, the usual convention is to give header files names that end with `.h'.
+It is most portable to use only letters, digits, dashes, and underscores in
+header file names, and at most one dot.
+
+Read more about using header files in official GCC documentation:
```
Comment (m-mcgowan) [inline 18-temperature-and-humidity-monitor/firmware/include/README:32]:
```suggestion
Read more about using header files in the official GCC documentation:
```

File: 18-temperature-and-humidity-monitor/firmware/include/README [hunk 3]
```diff
@@ -0,0 +1,39 @@
+
+This directory is intended for project header files.
+
+A header file is a file containing C declarations and macro definitions
+to be shared between several project source files. You request the use of a
+header file in your project source file (C, C++, etc) located in `src` folder
```
Comment (m-mcgowan) [inline 18-temperature-and-humidity-monitor/firmware/include/README:6]:
```suggestion
header file in your project source file (C, C++, etc) located in the `src` folder
```

File: 18-temperature-and-humidity-monitor/firmware/include/README [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+by including it, with the C preprocessing directive `#include'.
+
+```src/main.c
+
+#include "header.h"
```
Comment (m-mcgowan) [inline 18-temperature-and-humidity-monitor/firmware/include/README:11]: lol I guess I shouldn't be correcting Platformio's prose.

File: 18-temperature-and-humidity-monitor/firmware/lib/README [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+#include <Bar.h>
+
+int main (void)
+{
+  ...
+}
+
+```
+
+PlatformIO Library Dependency Finder will find automatically dependent
```
Comment (m-mcgowan) [inline 18-temperature-and-humidity-monitor/firmware/lib/README:42]:
```suggestion
The PlatformIO Library Dependency Finder will automatically find dependent
```

File: 18-temperature-and-humidity-monitor/firmware/src/main.cpp [hunk 6]
```diff
@@ -0,0 +1,423 @@
...
+//
+
+#include <Notecard.h>
+#include <Wire.h>
+#include <SparkFunBME280.h>
+
+// Uncomment this line and replace com.your-company:your-product-name with your
+// ProductUID.
+// #define PRODUCT_UID "com.your-company:your-product-name"
+#define PRODUCT_UID "com.blues.hroche:nf18"
```
Comment (m-mcgowan) [inline 18-temperature-and-humidity-monitor/firmware/src/main.cpp:14]: you probably want to remove this.

Here's a link to the updated message in the pragma warning.

https://github.com/blues/app-accelerators/blob/main/08-power-monitoring-and-control/firmware/src/app.h#L14

File: 18-temperature-and-humidity-monitor/firmware/src/main.cpp [hunk 7]
```diff
@@ -0,0 +1,423 @@
...
+    state.monitorIntervalMs = MONITOR_INTERVAL_DEFAULT_MS;
+    state.temperatureMin = TEMPERATURE_MIN_DEFAULT;
+    state.temperatureMax = TEMPERATURE_MAX_DEFAULT;
+    state.humidityMin = HUMIDITY_MIN_DEFAULT;
+    state.humidityMax = HUMIDITY_MAX_DEFAULT;
+
+    // Configure the productUID and instruct the Notecard to sync outbound notes
+    // a minimum of every OUTBOUND_SYNC_MINS minutes.
+    J *req = notecard.newRequest("hub.set");
+    JAddStringToObject(req, "product", PRODUCT_UID);
```
Comment (m-mcgowan) [inline 18-temperature-and-humidity-monitor/firmware/src/main.cpp:292]: `PRODUCT_UID` may be the empty string, so best to make this conditional

```c++
if (PRODUCT_UID[0]) {
   JAddStringToObject(req, "product", PRODUCT_UID);
}
```
(I noticed I didn't in NF8, so will also fix that one up.)

File: 18-temperature-and-humidity-monitor/firmware/src/main.cpp [hunk 8]
```diff
@@ -0,0 +1,423 @@
...
+
+    if (state.humidity < state.humidityMin) {
+        humidityStatus = statusLow;
+    }
+    else if (state.humidity > state.humidityMax) {
+        humidityStatus = statusHigh;
+    }
+
+    // If either status is not "ok".
+    if (strcmp(temperatureStatus, "ok") || strcmp(humidityStatus, "ok")) {
```
Comment (m-mcgowan) [inline 18-temperature-and-humidity-monitor/firmware/src/main.cpp:393]: nitpick: `statusOk` rather than repeating the string

**Review by m-mcgowan — CHANGES_REQUESTED** [review]
Looks good. just the PRODUCT_UID and the message is the main change, others just suggestive.

File: 18-temperature-and-humidity-monitor/firmware/README.md (hunk 1 above)
Comment (haydenroche5) [reply 18-temperature-and-humidity-monitor/firmware/README.md:76]: I'm an e.g. fan. I'll change to that.

File: 18-temperature-and-humidity-monitor/firmware/src/main.cpp (hunk 6 above)
Comment (haydenroche5) [reply 18-temperature-and-humidity-monitor/firmware/src/main.cpp:14]: Updated the wording but I'd like to keep it as `#error`. I don't see any point in continuing with the build if PRODUCT_UID isn't defined.

Comment (m-mcgowan) [reply 18-temperature-and-humidity-monitor/firmware/src/main.cpp]: We make it a warning so the user may choose to configure the productUID as part of the notecard setup, as we've done in other projects.

**Review by m-mcgowan — CHANGES_REQUESTED** [review]

**Review by m-mcgowan — APPROVED** [review]

**Pull request merged**